src/main/java/info/kgeorgiy/java/advanced/student/StudentDB.java

```diff
@@ -0,0 +1,169 @@
+    public List<String> getFirstNames(List<Student> students) {
+        return Strings.list(of(students, Student::getFirstName));
+    }
+
+    @Override
+    public List<String> getLastNames(List<Student> students) {
+        return Strings.list(of(students, Student::getLastName));
+    }
+
+    @Override
+    public List<String> getGroups(List<Student> students) {
+        return Strings.list(of(students, Student::getGroup));
+    }
+
+    @Override
+    public List<String> getFullNames(List<Student> students) {
+        return Strings.list(of(students, this::getFullName));
+    }
+
+    @Override
+    public Set<String> getDistinctFirstNames(List<Student> students) {
+        return Strings.set(of(students, Student::getFirstName));
+    }
+
+    @Override
+    public String getMinStudentFirstName(List<Student> students) {
+        return min(of(students))
+                .map(Student::getFirstName)
+                .orElse(DEFAULT_EMPTY_STRING);
+    }
+
+    @Override
+    public List<Student> sortStudentsById(Collection<Student> students) {
+        return Students.list(sorted(of(students)));
+    }
+
+    @Override
+    public List<Student> sortStudentsByName(Collection<Student> students) {
+        return Students.list(
+                sorted(
+                        of(students),
+                        with(Compare.BY_NAME)));
+    }
+
+    @Override
+    public List<Student> findStudentsByFirstName(Collection<Student> students, String name) {
+        return Students.list(
+                sorted(
+                        filtered(of(students),
+                                Student::getFirstName, name),
+                        with(Compare.BY_NAME)));
+    }
+
+    @Override
+    public List<Student> findStudentsByLastName(Collection<Student> students, String name) {
+        return Students.list(
+                sorted(
+                    filtered(of(students),
+                            Student::getLastName, name),
+                        with(Compare.BY_NAME)));
+    }
+
+    @Override
+    public List<Student> findStudentsByGroup(Collection<Student> students, String group) {
+        return Students.list(
+                sorted(
+                        filtered(of(students),
+                                Student::getGroup, group),
+                        with(Compare.BY_NAME)));
+    }
+
+    @Override
+    public Map<String, String> findStudentNamesByGroup(Collection<Student> students, String group) {
+        return mapOf(
+                filtered(
+                        of(students),
+                        Student::getGroup, group),
+                Student::getLastName,
+                Student::getFirstName,
+                BinaryOperator.minBy(Comparable::compareTo));
+    }
+
+    //region Helper Collector Stuff
+
+    /**
+     * Had to do all of it since I can't just make beautiful Stream to List type safe converted method
+     *
+     * @param <T>
+     */
+    private static class Helper<T> {
```

Comment:
I didn't get why do you need this. you can create a static method like `private static <T> List<T> list(Stream<T> stream)`

Reply:
wow, can I? I thought that you can't create generic methods in non-generic classes... 

Oh, I see, I was missing leading `<T>`, awkward, but awesome, I'm so fixing this